test/e2e/keycloak_client_test.go

```diff
@@ -104,3 +109,67 @@ func keycloakClientBasicTest(t *testing.T, framework *test.Framework, ctx *test.
 func externalKeycloakClientBasicTest(t *testing.T, framework *test.Framework, ctx *test.Context, namespace string) error {
 	return WaitForClientToBeReady(t, framework, namespace, testExternalKeycloakClientCRName)
 }
+
+func getKeycloakClientWithRolesCR(namespace string, roleNames []string) *keycloakv1alpha1.KeycloakClient {
+	client := getKeycloakClientCR(namespace, false)
+	client.Spec.Roles = getKeycloakClientRoles(roleNames)
+	return client
+}
+
+func getKeycloakClientRoles(roleNames []string) []keycloakv1alpha1.RoleRepresentation {
+	var roles []keycloakv1alpha1.RoleRepresentation
+	for _, roleName := range roleNames {
+		roles = append(roles, keycloakv1alpha1.RoleRepresentation{
+			Id: roleName,
+			Name: roleName,
+		})
+	}
+	return roles
+}
+
+func keycloakClientRolesTest(t *testing.T, framework *test.Framework, ctx *test.Context, namespace string) error {
+	// create
+	err := Create(framework, getKeycloakClientWithRolesCR(namespace, []string{"a", "b"}), ctx)
+	if err != nil {
+		return err
+	}
+	err = WaitForClientToBeReady(t, framework, namespace, testKeycloakClientCRName)
+	if err != nil {
+		return err
+	}
+
+	// update client: delete/leave/add role
+	var client keycloakv1alpha1.KeycloakClient
+	err = GetNamespacedObject(framework, namespace, testKeycloakClientCRName, &client)
+	if err != nil {
+		return err
+	}
+	client.Spec.Roles = getKeycloakClientRoles([]string{"b", "c"})
+	err = Update(framework, &client)
+	if err != nil {
+		return err
+	}
+	err = ensureOperatorRan(t, framework, ctx, namespace)
```

Comment:
I'm not sure I'm happy with this approach. It feels to me a bit nasty. Could we rather somehow check if the roles were truly updated in the CR?

Reply:
The closest I got was looking into `managedFields` with `manager: keycloak-operator` but while it seems to be a better approach to look in the CR than to "implement the barrier" in this way I did not find information on how to use this in a reliable and clear way.
The alternative of manually setting status to something the operator will change again did not work either since `.status` cannot simply be updated.
The other tests just wait for the status but for testing a true update this is simply not enough.
It is a workaround but unfortunately the most reliable and clear way I could find so far that does not include accessing the KC API directly from the test project (then it could just check for the expected changes).

Reply:
AFAIK, it should be possible to retrieve current CR. Please see [`GetNamespacedObject`](https://github.com/keycloak/keycloak-operator/blob/e655f80a246509ff13bd4d607a40fc77bd2a004f/test/e2e/utils.go#L213) usages.

Reply:
Yes that is undoubtedly possible. What approach are you suggesting? Do you have information that would suggest the `managedFields` approach is reliable and portable enough to not break from slight changes/updates to the e2e test environment?

Reply:
I'd just fetch a fresh client object using `GetNamespacedObject` and check the roles are really there. That should be robust enough. WDYT?

Reply:
The roles are in the k8s object the instant `Update()` (L148) completes.
The test should ensure the operator was actually able to perform the necessary changes on the KC server and I think this can only be verified by either checking on the KC server (which is not done in any other e2e test AFAICS) or checking for success in the CR after the operator tried to reconcile (which is what's happening right now).
The tricky part is making sure the operator actually tried to reconcile and we're not just going with the previous `phase: Reconciling` because the operator was too slow / e2e tests too fast.
Other e2e tests don't have this problem as they don't try to update from a successfully reconciled state.

Reply:
Ok, I understand. I think `ensureOperatorRan` should be at least placed into more generic [`utils`](https://github.com/keycloak/keycloak-operator/blob/e655f80a246509ff13bd4d607a40fc77bd2a004f/test/e2e/utils.go) in case some other test needs it in the future.

Reply:
Turns out directly asking the KC server is easier than expected. It also seems cleaner. See updated commit.